internal/witness/witness.go

```diff
@@ -114,88 +119,94 @@ func New(wo Opts) (*Witness, error) {
 // the parsed checkpoint and the note itself.
-func (w *Witness) parse(chkptRaw []byte, logID string) (*log.Checkpoint, *note.Note, error) {
+func (w *Witness) parse(chkptRaw []byte) (*log.Checkpoint, *note.Note, LogInfo, error) {
+	origin, _, found := strings.Cut(string(chkptRaw), "\n")
+	if !found {
+		return nil, nil, LogInfo{}, errors.New("invalid checkpoint")
+	}
+	logID := log.ID(origin)
 	logInfo, ok := w.Logs[logID]
 	if !ok {
-		return nil, nil, ErrUnknownLog
+		return nil, nil, LogInfo{}, ErrUnknownLog
 	}
 	cp, _, n, err := log.ParseCheckpoint(chkptRaw, logInfo.Origin, logInfo.SigV)
-	return cp, n, err
-}
-
-// GetLogs returns a list of all logs the witness is aware of.
-func (w *Witness) GetLogs() ([]string, error) {
-	return w.lsp.Logs()
+	return cp, n, logInfo, err
 }
 
 // GetCheckpoint gets a checkpoint for a given log, which is consistent with all
 // other checkpoints for the same log signed by this witness.
-func (w *Witness) GetCheckpoint(logID string) ([]byte, error) {
-	chkpt, err := w.lsp.Latest(logID)
-	if err != nil {
-		return nil, err
-	}
-	return chkpt, nil
+//
+// Returns a nil checkpoint if no checkpoint is stored for the given logID.
+func (w *Witness) GetCheckpoint(_ context.Context, logID string) ([]byte, error) {
+	return w.lsp.Latest(logID)
 }
 
 // Update updates the latest checkpoint if nextRaw is consistent with the current
 // latest one for this log.
 //
-// It returns the latest cosigned checkpoint held by the witness (which may be the provided checkpoint
-// if it's accepted).
-func (w *Witness) Update(ctx context.Context, logID string, oldSize uint64, nextRaw []byte, cProof [][]byte) ([]byte, error) {
-	// If we don't witness this log then no point in going further.
-	logInfo, ok := w.Logs[logID]
-	if !ok {
-		return nil, ErrUnknownLog
-	}
-	counterUpdateAttempt.Inc(logID)
+// The values returned depend on whether or not the new checkpoint is accepted, and
+// if not, the reason it was rejected. This can be determined through the error:
+//
+// - no error: The checkpoint was accepted, and a serialised note-signature is returned.
+// - ErrCheckpointStale or ErrOldSizeInvalid: the presented checkpoint is out of date, the size of the current checkpoint is returned.
+// - Any other error, no supporting values are returned.
+func (w *Witness) Update(ctx context.Context, oldSize uint64, nextRaw []byte, cProof [][]byte) ([]byte, uint64, error) {
 	// Check the signatures on the raw checkpoint and parse it
 	// into the log.Checkpoint format.
 	//
 	// SPEC: The witness MUST verify the checkpoint signature against the public key(s) it trusts for the
 	//       checkpoint origin, and it MUST ignore signatures from unknown keys.
-	next, nextNote, err := w.parse(nextRaw, logID)
+	next, nextNote, logInfo, err := w.parse(nextRaw)
 	if err != nil {
 		// TODO(al): Technically this could also be that the checkpoint body is invalid.
-		return nil, ErrNoValidSignature
+		return nil, 0, ErrNoValidSignature
```

Comment:
Converting the `ErrUnknownLog` to `ErrNoValidSignature` feels a bit of a stretch... I assume it's a mistake? From the spec: `If the checkpoint origin is unknown, the witness MUST respond with a "404 Not Found" HTTP status code.`. This conversion from `ErrUnknownLog` to 404 is performed, but can't work if the error is replaced like this.